File: HOT100/8.2/543.二叉树的直径.py
#
# @lc app=leetcode.cn id=543 lang=python3
#
# [543] 二叉树的直径
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 1
        def dfs(n: Optional[TreeNode]):
            nonlocal res
            if not n: return 0
            l = dfs(n.left)
            r = dfs(n.right)
            res = max(res, l + r + 1)
            return max(l, r) + 1
        dfs(root)
        return res - 1
# @lc code=end

File: HOT100/HOT100p1/8.二叉树/543.二叉树的直径.cpp
/*
 * @lc app=leetcode.cn id=543 lang=cpp
 *
 * [543] 二叉树的直径
 */
#include "tools.h"
using namespace std;
struct TreeNode
{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 0;
    int diameterOfBinaryTree(TreeNode* root) {
        
        auto dfs = [&](auto&& self, TreeNode* n)
        {
            if (! n) return 0;
            int l = self(self, n->left);
            int r = self(self, n->right);
            res = max(res, l + r);
            return max(l, r) + 1;
        };
        dfs(dfs, root);
        return res;
    }
};
// @lc code=end

